subsystems/networking/networkmanager_linux.go

```diff
@@ -823,12 +825,18 @@ func (n *Networking) mainLoop(ctx context.Context) {
 			shouldReboot := n.Config().DeviceRebootAfterOfflineMinutes > 0 &&
 				lastConnectivity.Before(now.Add(time.Duration(n.Config().DeviceRebootAfterOfflineMinutes)*-1))
 
-			shouldExit := allGood || haveCandidates || fallbackHit || shouldReboot || userInputReceived
+			shouldExitPMode := (allGood || haveCandidates || fallbackHit || shouldReboot || userInputReceived) &&
+				!forceStayInPModeUntilRecvUserInput
 
 			n.logger.Debugf("inactive portal: %t, have candidates: %t, fallback timeout: %t (%s remaining)",
 				inactivePortal, haveCandidates, fallbackHit, fallbackRemaining)
 
-			if shouldExit {
+			if shouldExitPMode {
+				if userInputReceived {
+					// We could get to this point before the user receives our response (poor UX, but likely not critical)
+					// E.g. try to avoid "Not connected" web portal screen.
+					time.Sleep(3 * time.Second)
```

Comment:
You should never use regular sleep in agent (or in general in most of Viam's code.) Here, there's a "Sleep()" method on the two healthcheck objects for background and foreground loops that both keeps the healthcheck satisfied and obeys context cancellation. This is the main loop, so probably `n.mainLoopHealth.Sleep(ctx, 3 * time.Second)`

That said, I don't understand why this is needed at all, when it wasn't needed before. I'm pretty sure that the return to any GRPC calls should've already happened before userInputReceived would've been set true.